tests/test_metrics.py

```diff
@@ -29,76 +29,95 @@ class Meta:
-    assert (
-        PreprintView.get_index_name(date=date) == "osf_metrics_preprintviews-2020.02.14"
-    )
-
-
-def test_get_index_name_gets_index_for_today_by_default():
-    today = timezone.now().date()
-    # TODO: should be in settings
-    dateformat = "%Y.%m.%d"
-    today_formatted = today.strftime(dateformat)
-    assert PreprintView.get_index_name() == "osf_metrics_preprintviews-{}".format(
-        today_formatted
-    )
-
-
-@pytest.mark.es
-def test_create_metric_creates_template_with_mapping(client):
-    PreprintView.create_index_template()
-    template_name = PreprintView._template_name
-    template = client.indices.get_template(name=template_name)
-    mappings = template[template_name]["mappings"]
-    assert mappings["doc"]["_all"]["enabled"] is False
-    assert mappings["doc"]["_source"]["enabled"] is False
-    properties = mappings["doc"]["properties"]
-    assert "timestamp" in properties
-    assert properties["timestamp"] == {"doc_values": True, "type": "date"}
-    assert properties["provider_id"] == {"type": "keyword", "index": True}
-    assert properties["user_id"] == {"type": "keyword", "index": True}
-    assert properties["preprint_id"] == {"type": "keyword", "index": True}
-
-
-def test_create_metric_sends_pre_index_template_create_signal():
-    mock_listener = mock.Mock()
-    pre_index_template_create.connect(mock_listener)
-    PreprintView.create_index_template()
-    assert mock_listener.call_count == 1
-    assert "index_template" in mock_listener.call_args[1]
-
-
-# TODO: Can we make this test not use ES?
-@pytest.mark.es
-def test_source_may_be_enabled(client):
-    class MyMetric(Metric):
-        class Meta:
-            template_name = "mymetric"
-            template = "mymetric-*"
-            source = MetaField(enabled=True)
-
-    MyMetric.create_index_template()
-    template_name = MyMetric._template_name
-    template = client.indices.get_template(name=template_name)
-    mappings = template[template_name]["mappings"]
-    assert mappings["doc"]["_all"]["enabled"] is False
-    assert mappings["doc"]["_source"]["enabled"] is True
+class TestPreprintView:
+    @classmethod
+    def setup_class(cls):
+        # TODO hook into pytest.mark.es to delete indices
+        client().indices.delete(index="*")
+
+    @classmethod
+    def teardown_class(cls):
+        client().indices.delete(index="*")
+
+    def test_get_index_template(self):
+        template = PreprintView.get_index_template()
+        assert isinstance(template, IndexTemplate)
+        assert template._template_name == "osf_metrics_preprintviews"
+        assert "osf_metrics_preprintviews-*" in template.to_dict()["index_patterns"]
+
+    @pytest.mark.es
+    def test_create_index(self, client):
+        PreprintView.create_index()
+        name = PreprintView.get_index_name()
+        mapping = client.indices.get_mapping(index=name)
+        properties = mapping[name]["mappings"]["doc"]["properties"]
```

Comment:
This is passing locally..hmm

Reply:
Ah, woops this might be a timing issue. I set `refresh_interval` to `5s`, which may have to do with this. Maybe try setting it to `-1`.

Reply:
I was able to reproduce the error with a fresh ES cluster (restarted the ES container). After the first run, it passes.

If you clear out the index templates before and after each run, you can consistently reporduce it:

```diff
diff --git a/tests/test_metrics.py b/tests/test_metrics.py
index e7f94f6..a24a308 100644
--- a/tests/test_metrics.py
+++ b/tests/test_metrics.py
@@ -21,7 +21,7 @@ class PreprintView(Metric):
     preprint_id = Keyword(index=True)
 
     class Index:
-        settings = {"refresh_interval": "5s"}
+        settings = {"refresh_interval": "-1"}
 
     class Meta:
         # TODO: Make this unnecessary and compute this.
@@ -34,10 +34,12 @@ class TestPreprintView:
     def setup_class(cls):
         # TODO hook into pytest.mark.es to delete indices
         client().indices.delete(index="*")
+        client().indices.delete_template("*")
 
     @classmethod
     def teardown_class(cls):
         client().indices.delete(index="*")
+        client().indices.delete_template("*")
 
     def test_get_index_template(self):
         template = PreprintView.get_index_template()
```

Reply:
Oh, the error is expected because the test isn't creating the index template before creating the index, so the mapping won't be there.

This raises some important questions: 

1. Should we allow indices to be created without the template being created? If not, then should we a) implicitly create the index template? b) go ahead and create the index *with* the mapping? c) raise an error?
2. Do we actually need the index templates at all? If we're defining the mappings on the `Metric` classes anyway, then we're already enforcing a correct mapping.

I lean towards 1b (create the index with the correct mapping without creating the index template) because I think it's the most intuitive.

I'm unsure about 2. It might still be nice to have a template mapping, in case users create indices outside of this library.

@pattisdr Thoughts?

Reply:
ah, i see. an early version of my test was creating the template before creating the index, but I was getting errors that the template was already there. I forgot that I needed to be clearing out the templates too.

Reply:
I wouldn't expect `create_index` to create an index template for the same reason I wouldn't expect `Model.save()` to create a table. It seems outside its responsibility. It's the user's responsibility to ensure that `manage.py sync_metrics` (which will call `create_index_template`) is called before new indices are created (like it's their responsibility to call `manage.py migrate`).

Also, to be clear: I think that `create_index` should still respect the mapping defined by the class--I just don't think it should also create an index template.

Reply:
okay, that's fine.